### PR TITLE
HIVE-27352: Support both LDAP and Kerberos Auth in HS2

### PR DESCRIPTION
**Pull request by dengzhhu653**
### What changes were proposed in this pull request?
Support both LDAP and Kerberos Auth in HS2

### Does this PR introduce _any_ user-facing change?
Yes. For Http mode, we must append the new auth method to the end of property hive.server2.authentication in order to be compatible with the old HS2 client.

### How was this patch tested?
Unit tests

### CODE REVIEW

File: itests/hive-minikdc/src/test/java/org/apache/hive/minikdc/TestHS2JWTWithMiniKdc.java [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+
+  @BeforeClass
+  public static void setUpBeforeClass() throws Exception {
+    Class.forName(MiniHS2.getJdbcDriverName());
+    MOCK_JWKS_SERVER.stubFor(get("/jwks")
+        .willReturn(ok()
+            .withBody(Files.readAllBytes(jwtVerificationJWKSFile.toPath()))));
+    miniHiveKdc = new MiniHiveKdc();
+    HiveConf hiveConf = new HiveConf();
+    hiveConf.setBoolVar(HiveConf.ConfVars.HIVE_SUPPORT_CONCURRENCY, false);
```
Comment (saihemanth-cloudera) [inline itests/hive-minikdc/src/test/java/org/apache/hive/minikdc/TestHS2JWTWithMiniKdc.java:82]: This is anyway set to false by default.

File: service/src/java/org/apache/hive/service/auth/AuthType.java [hunk 2]
```diff
@@ -18,55 +18,78 @@
...
-      return;
+    if (typeBits.get(HiveAuthConstants.AuthTypes.NOSASL.ordinal())) {
+      throw new UnsupportedOperationException("NOSASL can't be along with other auth methods: " + authTypes);
+    }
+
+    if (typeBits.get(HiveAuthConstants.AuthTypes.NONE.ordinal())) {
+      throw new UnsupportedOperationException("None can't be along with other auth methods: " + authTypes);
+    }
+
+    if (areAnyEnabled(PASSWORD_BASED_TYPES) && !isExactlyOneEnabled(PASSWORD_BASED_TYPES)) {
```
Comment (saihemanth-cloudera) [inline service/src/java/org/apache/hive/service/auth/AuthType.java:84]: Shouldn't we have multiple password-based types at once?

Comment (dengzhhu653) [reply service/src/java/org/apache/hive/service/auth/AuthType.java:84]: It's hard to tell which password-based type the client is using, as in HiveConnection we don't care about the specific type, instead it transmits the username/password directly to the server.
binary: https://github.com/apache/hive/blob/master/jdbc/src/java/org/apache/hive/jdbc/HiveConnection.java#L982-L987
http: https://github.com/apache/hive/blob/master/jdbc/src/java/org/apache/hive/jdbc/HiveConnection.java#L620-L627

File: itests/hive-minikdc/pom.xml [hunk 3]
```diff
@@ -29,6 +29,30 @@
   </properties>
   <dependencies>
     <!-- dependencies are always listed in sorted order by groupId, artifactId -->
+    <dependency>
+      <groupId>com.github.tomakehurst</groupId>
+      <artifactId>wiremock-jre8-standalone</artifactId>
+      <version>2.32.0</version>
```
[the remark this answers is not in the document]
Comment (dengzhhu653) [reply itests/hive-minikdc/pom.xml]: ack

File: itests/hive-minikdc/src/test/java/org/apache/hive/minikdc/TestHS2JWTWithMiniKdc.java (hunk 1 above)
Comment (dengzhhu653) [reply itests/hive-minikdc/src/test/java/org/apache/hive/minikdc/TestHS2JWTWithMiniKdc.java]: ack

Comment (dengzhhu653) [reply itests/hive-minikdc/src/test/java/org/apache/hive/minikdc/TestHS2JWTWithMiniKdc.java]: Some tests failed, after some investigation, The test will read the properties from:
https://github.com/apache/hive/blob/master/data/conf/hive-site.xml#L183-L187

Restore the HiveConf.ConfVars.HIVE_SUPPORT_CONCURRENCY to false.

**Review by saihemanth-cloudera — COMMENTED** [review]
Mostly looks good, I will approve once the clarification on the review comment is resolved.

File: service/src/test/org/apache/hive/service/auth/TestAuthType.java [hunk 4]
```diff
@@ -75,10 +79,40 @@ private void testOnePasswordAuthWithSAML(HiveAuthConstants.AuthTypes type) throw
       Assert.assertFalse(authType.isEnabled(disabledType));
     }
     Assert.assertEquals(type.getAuthName(), authType.getPasswordBasedAuthStr());
+
+    verify("SAML," + type.getAuthName(), HiveServer2TransportMode.binary, true);
+    verify("SAML," + type.getAuthName(), HiveServer2TransportMode.all, true);
```
[the remark this answers is not in the document]
Comment (dengzhhu653) [reply service/src/test/org/apache/hive/service/auth/TestAuthType.java:84]: In all mode, HS2 will start both binary and http thrift service,  and currently the only property: hive.server2.authentication works for both modes, but the binary mode doesn't support JWT and SAML yet, so in all mode, these haven't been supported.

**Pull request merged**